### DISCUSSION TITLE
Platform-specific behavior in FastAPI dependency injection during async tests

### DISCUSSION THREAD

**Discussion by weijay0804** [fastapi/fastapi#12183] [Questions]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
@company_router.patch(
    "/{company_id}",
    response_model=ReadCompany,
    dependencies=[Depends(PermissionDependency([AdminPermission]))],
)
async def update_company(
    data: UpdateCompany,
    company_id: PydanticObjectId = Depends(get_oid_or_400("company_id")),
):

    company = await company_curd.get(company_id)

    return await company_curd.update(db_obj=company, obj_in=data)
```

### Description

I have a route that involves permission checks.
The PermissionDependency will check the provided BasePermission instance and verify if the user has the required permissions. If not, it will raise a PermissionDeniedError.

```python
class PermissionDependency:

    def __init__(self, permissions_classes: List["BasePermission"]):
        self.permissions_classes = permissions_classes

    async def __call__(self, request: Request, current_user: CurrentActiveUser):

        for permission_class in self.permissions_classes:

            p = permission_class(request, current_user)

            await p.initialize()
```

`BasePermission class`

```python
class BasePermission(ABC):

    @abstractmethod
    def has_request_permission(self) -> bool: ...  # noqa: E704

    def __init__(self, request: Request, current_user: CurrentActiveUser):

        self.current_user = current_user
        self.request = request
        self.role = None

    async def initialize(self):
        if self.current_user.is_superuser:
            return

        user_company = None

        if self.request.path_params.get("company_id"):

            company_id = validate_oid(self.request.path_params["company_id"])
            user_id = validate_oid(self.current_user.id)

            user_company = await UserCompany.find_one(
                UserCompany.user.id == user_id, UserCompany.company.id == company_id
            )

        if not user_company:
            raise EntityNotExistError()

        self.role = user_company.role

        if not await self.has_request_permission():
            raise PermissionDeniedError()
```

unit test for this route

```python
@pytest.mark.asyncio
async def test_update_company_permission_denied(client, mocker: MockFixture, mock_member):

    _, company, headers = mock_member

    update_data = {"name": "Test Company Updated"}


    mocker.patch(
        "app.company.router.PermissionDependency.__call__", side_effect=PermissionDeniedError
    )

    response = await client.patch(f"/company/{company.id}", headers=headers, json=update_data)

    assert response.status_code == 403
```

When I run the tests on macOS, they pass successfully.
![截圖 2024-09-11 中午12 00 38](https://github.com/user-attachments/assets/6ef53fbe-f625-4f46-b98c-af2acfbd97cf)

when I run them on Windows, they do not pass.
![___________________2024-09-11_120523_720](https://github.com/user-attachments/assets/bfc8d74c-0268-4c10-8b3e-5c5ddb0fbbd7)

If you provide the updated version of the route
```python
@company_router.patch(
    "/{company_id}",
    response_model=ReadCompany,
    # dependencies=[Depends(PermissionDependency([AdminPermission]))],
)
async def update_company(
    data: UpdateCompany,
    test = Depends(PermissionDependency([AdminPermission])),
    company_id: PydanticObjectId = Depends(get_oid_or_400("company_id")),
):

    await test

    company = await company_curd.get(company_id)

    return await company_curd.update(db_obj=company, obj_in=data)
```

they pass successfully.
![a](https://github.com/user-attachments/assets/0bac2cd5-2792-4732-a7b6-8d4c885ec1bf)

I’m not sure if the difference is due to the underlying event loop mechanisms on Windows and macOS.

### Operating System

Windows, macOS

### Operating System Details

MacOS:
- Mac mini M1,
- Sonoma 14.4.1

Window:
- Window 11

### FastAPI Version

0.112.1

### Pydantic Version

2.8.2 

### Python Version

3.10

### Additional Context

- `pytest` version: 8.3.2
- `pytest-asyncio` version: 0.24.0
- `pytest-mock` version: 3.14.0

The environments on Windows and macOS are the same.

**Comment by Jkrox** [comment]
can you show more details about pytest output?

**Reply by weijay0804** [reply]
here are details
```bash
____________________________________________________________________________ test_update_company_permission_denied ____________________________________________________________________________ 

client = <httpx.AsyncClient object at 0x000001F72F8920B0>, mocker = <pytest_mock.plugin.MockerFixture object at 0x000001F72F893220>
mock_member = (User(id=ObjectId('66e29a1464df5b0a3c72cfaa'), revision_id=None, email='member@test.com', hashed_password='member', is...3MmNmYWEiLCJhdWQiOlsiZmFzdGFwaS11c2VyczphdXRoIl0sImV4cCI6MTcyNjEzMDIxMn0.bjXnlhp2f6U9QzqTcTH9wdHdH5PozM89iYiKYhQ2Ym4'})

    @pytest.mark.asyncio
    async def test_update_company_permission_denied(client, mocker: MockFixture, mock_member):

        _, company, headers = mock_member

        update_data = {"name": "Test Company Updated"}

        mocker.patch(
            "app.company.router.PermissionDependency.__call__", side_effect=PermissionDeniedError
        )

        response = await client.patch(f"/company/{company.id}", headers=headers, json=update_data)

>       assert response.status_code == 403
E       assert 200 == 403
E        +  where 200 = <Response [200 OK]>.status_code

tests\company\test_company_router.py:176: AssertionError
=================================================================================== short test summary info =================================================================================== 
FAILED tests/company/test_company_router.py::test_update_company_permission_denied - assert 200 == 403
```

Because the `PermissionDeniedError` was not triggered, it returns a status code of 200.

Here is the `any_permission` function and the `AdminPermission` class, both of which are related to the previously mentioned `BasePermission`.

`any_permission` function:
```python
async def any_permission(
    permissions: list, request: Request, current_user: CurrentActiveUser
) -> bool:

    for p in permissions:

        try:

            p_obj = p(request, current_user)

            await p_obj.initialize()
            return True

        except PermissionDeniedError:
            pass

    return False
```

`AdminPermission` class:
```python
class AdminPermission(BasePermission):

    async def has_request_permission(self) -> bool:

        permission = await any_permission(
            permissions=[SuperUserPermission], request=self.request, current_user=self.current_user
        )

        if not permission:

            if self.role == Role.admin.value:
                return True

        return permission
```

**Reply by Jkrox** [reply]
Hi, 

Sorry for not reply your question, I was wondering if you was able to solve it, I was about to ask you for more context of your code to try to debug and see what's going on under the hood, I was thinking that something happens weird with the `MockFixture` -> `"app.company.router.PermissionDependency.__call__", side_effect=PermissionDeniedError`, even though not pretty sure.

**Comment by YuriiMotov** [comment]
Let's close this for now to keep the discussions clean.
If you’ve found a solution, please share it here.
If you still need help, feel free to let us know.
